### PR TITLE
Start consume from specific offset

### PR DESCRIPTION
**Pull request by shohamroditimemphis**

### CODE REVIEW

File: models/consumers.go [hunk 1]
```diff
@@ -32,6 +32,8 @@ type Consumer struct {
 	CreationDate     time.Time          `json:"creation_date" bson:"creation_date"`
 	IsDeleted        bool               `json:"is_deleted" bson:"is_deleted"`
 	MaxMsgDeliveries int                `json:"max_msg_deliveries" bson:"max_msg_deliveries"`
+	OptStartSequence uint64             `json:"opt_start_sequence"`
```
Comment (idanasulinmemphis) [inline models/consumers.go:35]: we decided to use different name

File: server/memphis_handlers_consumers.go [hunk 2]
```diff
@@ -246,10 +246,13 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
 		IsDeleted:        false,
 		MaxAckTimeMs:     int64(ccr.MaxAckTimeMillis),
 		MaxMsgDeliveries: ccr.MaxMsgDeliveries,
+		OptStartSequence: ccr.OptStartSequence,
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:249]: you should validate you didn't get both of them

File: server/memphis_handlers_consumers.go [hunk 3]
```diff
@@ -246,10 +246,13 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
...
+		LastMessages:     ccr.LastMessages,
 	}
 
 	if consumerGroupExist {
-		if newConsumer.MaxAckTimeMs != consumerFromGroup.MaxAckTimeMs || newConsumer.MaxMsgDeliveries != consumerFromGroup.MaxMsgDeliveries {
+
+		if newConsumer.MaxAckTimeMs != consumerFromGroup.MaxAckTimeMs || newConsumer.MaxMsgDeliveries != consumerFromGroup.MaxMsgDeliveries || newConsumer.OptStartSequence != consumerFromGroup.OptStartSequence || newConsumer.LastMessages != consumerFromGroup.LastMessages {
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:255]: in this case you are updating the current consumer and you said that we cant do that

File: server/memphis_helper.go [hunk 4]
```diff
@@ -361,19 +361,66 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
-		AckWait:       time.Duration(maxAckTimeMs) * time.Millisecond,
-		MaxDeliver:    MaxMsgDeliveries,
-		FilterSubject: stationName.Intern() + ".final",
-		ReplayPolicy:  ReplayInstant,
-		MaxAckPending: -1,
-		HeadersOnly:   false,
-		// RateLimit: ,// Bits per sec
-		// Heartbeat: // time.Duration,
-	})
+	var ackPolicy DeliverPolicy
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:364]: this is not ack policy it is a deliveryPolicy

File: server/memphis_helper.go [hunk 5]
```diff
@@ -361,19 +361,66 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
+			DeliverPolicy: ackPolicy,
+			AckPolicy:     AckExplicit,
+			AckWait:       time.Duration(maxAckTimeMs) * time.Millisecond,
+			MaxDeliver:    MaxMsgDeliveries,
+			FilterSubject: stationName.Intern() + ".final",
+			ReplayPolicy:  ReplayInstant,
+			MaxAckPending: -1,
+			HeadersOnly:   false,
+		}
+	} else if consumer.OptStartSequence != 0 {
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:406]: you should set it to 1 when it is < 1

File: server/memphis_helper.go [hunk 6]
```diff
@@ -361,19 +361,66 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
+	}
+	lastSeq := streamInfo.State.LastSeq
+
+	if consumer.LastMessages != 0 {
+		lastMessages := (lastSeq - consumer.LastMessages) + 1
+		if int(lastMessages) < 0 {
+			lastMessages = uint64(1)
+		}
+		ackPolicy = DeliverByStartSequence
+		consumerConfig = &ConsumerConfig{
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:378]: you don't need to re-write all the keys in each block. write the keys appears in all blocks and inject those which can be changed

File: server/memphis_sdk_handlers.go [hunk 7]
```diff
@@ -78,6 +78,8 @@ type createConsumerRequest struct {
 	ConsumerGroup    string `json:"consumers_group"`
 	MaxAckTimeMillis int    `json:"max_ack_time_ms"`
 	MaxMsgDeliveries int    `json:"max_msg_deliveries"`
+	OptStartSequence uint64 `json:"opt_start_sequence"`
```
Comment (idanasulinmemphis) [inline server/memphis_sdk_handlers.go:81]: change the name

File: server/memphis_handlers_consumers.go (hunk 3 above)
Comment (shohamroditimemphis) [reply server/memphis_handlers_consumers.go:255]: I did it from sdk side. I am adding it in the broker.

File: server/memphis_handlers_consumers.go [hunk 8]
```diff
@@ -233,22 +233,73 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
...
-		CreatedByUser:    connection.CreatedByUser,
-		ConsumersGroup:   consumerGroup,
-		IsActive:         true,
-		CreationDate:     time.Now(),
-		IsDeleted:        false,
-		MaxAckTimeMs:     int64(ccr.MaxAckTimeMillis),
-		MaxMsgDeliveries: ccr.MaxMsgDeliveries,
+	if ccr.StartConsumeFromSequence > 1 && ccr.LastMessages > -1 {
+		errMsg := "Consumer creation options can't contain both startConsumeFromSequence and lastMessages"
+		serv.Errorf("createConsumerDirect: " + errMsg)
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:238]: This is a warning not an error

File: server/memphis_handlers_consumers.go [hunk 9]
```diff
@@ -233,22 +233,73 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
...
+		respondWithErr(s, reply, err)
+		return
+	}
+
+	var newConsumer models.Consumer
+	if ccr.StartConsumeFromSequence > 1 {
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:244]: in each block you are re-write the entire parameters you can inject the relevant where needed

File: server/memphis_handlers_consumers.go [hunk 10]
```diff
@@ -233,22 +233,73 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
...
+			MaxAckTimeMs:             int64(ccr.MaxAckTimeMillis),
+			MaxMsgDeliveries:         ccr.MaxMsgDeliveries,
+			StartConsumeFromSequence: ccr.StartConsumeFromSequence,
+			LastMessages:             ccr.LastMessages,
+		}
 	}
 
 	if consumerGroupExist {
+		if newConsumer.StartConsumeFromSequence != consumerFromGroup.StartConsumeFromSequence || newConsumer.LastMessages != consumerFromGroup.LastMessages {
+			errMsg := errors.New("The consumer already exists with different configuration. You can't change the configuration to an existing consumer.")
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:297]: Consumer already exists with different uneditable configuration parameters (StartConsumeFromSequence/LastMessages)

File: server/memphis_handlers_consumers.go [hunk 11]
```diff
@@ -233,22 +233,73 @@ func (s *Server) createConsumerDirect(c *client, reply string, msg []byte) {
...
+			serv.Errorf("createConsumerDirect: " + errMsg.Error())
```
Comment (idanasulinmemphis) [inline server/memphis_handlers_consumers.go:298]: warn

File: server/memphis_helper.go [hunk 12]
```diff
@@ -361,19 +361,46 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
+	streamInfo, err := serv.memphisStreamInfo(stationName.Intern())
+	if err != nil {
+		return errors.New("Streaminfo: " + err.Error())
+	}
+	lastSeq := streamInfo.State.LastSeq
+
+	var optStartSeq uint64
+	if consumer.LastMessages > 0 {
+		lastMessages := (lastSeq - uint64(consumer.LastMessages)) + 1
+		if int(lastMessages) < 0 {
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:374]: < 1

File: server/memphis_helper.go [hunk 13]
```diff
@@ -361,19 +361,46 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
-	err = s.memphisAddConsumer(stationName.Intern(), &ConsumerConfig{
+	var deliveryPolicy DeliverPolicy
+	streamInfo, err := serv.memphisStreamInfo(stationName.Intern())
+	if err != nil {
+		return errors.New("Streaminfo: " + err.Error())
+	}
+	lastSeq := streamInfo.State.LastSeq
+
+	var optStartSeq uint64
+	if consumer.LastMessages > 0 {
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:372]: add if == 0 block

File: server/memphis_helper.go [hunk 14]
```diff
@@ -361,19 +361,46 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
-		DeliverPolicy: DeliverAll,
+		DeliverPolicy: deliveryPolicy,
 		AckPolicy:     AckExplicit,
 		AckWait:       time.Duration(maxAckTimeMs) * time.Millisecond,
 		MaxDeliver:    MaxMsgDeliveries,
 		FilterSubject: stationName.Intern() + ".final",
 		ReplayPolicy:  ReplayInstant,
 		MaxAckPending: -1,
 		HeadersOnly:   false,
+		OptStartSeq:   optStartSeq,
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:399]: add this parameter only when is not deliverAll policy

File: server/memphis_helper.go [hunk 15]
```diff
@@ -361,19 +361,46 @@ func (s *Server) CreateConsumer(consumer models.Consumer, station models.Station
...
+		lastMessages := (lastSeq - uint64(consumer.LastMessages)) + 1
+		if int(lastMessages) < 0 {
+			lastMessages = uint64(1)
+		}
+		deliveryPolicy = DeliverByStartSequence
+		optStartSeq = lastMessages
+	} else if consumer.StartConsumeFromSequence == 1 || consumer.LastMessages == -1 {
+		deliveryPolicy = DeliverAll
+	} else if consumer.StartConsumeFromSequence > 1 {
+		if int(consumer.StartConsumeFromSequence) < 0 {
```
Comment (idanasulinmemphis) [inline server/memphis_helper.go:382]: un reachable code

File: server/memphis_helper.go (hunk 13 above)
Comment (idanasulinmemphis) [reply server/memphis_helper.go:372]: in this case it is deliver new policy

**Pull request merged**